src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -158,106 +152,165 @@ private static final boolean isLocalAll(final String... requestedPatterns) {
             }
-            matchingAllIndices = Collections.emptySet();
-            matchingAliases = Collections.emptySet();
+
+            resolveTo(matchingAliases, matchingAllIndices, original, remoteIndices);
+
         }
 
-        else {
-            final ClusterState state = clusterService.state();
-            final Set<String> dateResolvedLocalRequestedPatterns = localRequestedPatterns
-                            .stream()
-                            .map(resolver::resolveDateMathExpression)
-                            .collect(Collectors.toSet());
-            final WildcardMatcher dateResolvedMatcher = WildcardMatcher.from(dateResolvedLocalRequestedPatterns);
-            //fill matchingAliases
-            final Map<String, IndexAbstraction> lookup = state.metadata().getIndicesLookup();
-            matchingAliases = lookup.entrySet()
-                    .stream()
-                    .filter(e -> e.getValue().getType() == ALIAS)
-                    .map(Map.Entry::getKey)
-                    .filter(dateResolvedMatcher)
-                    .collect(Collectors.toSet());
-
-            try {
-                matchingAllIndices = Arrays.asList(resolver.concreteIndexNames(state, indicesOptions, localRequestedPatterns.toArray(new String[0])));
-                if (log.isDebugEnabled()) {
-                    log.debug("Resolved pattern {} to {}", localRequestedPatterns, matchingAllIndices);
-                }
-            } catch (IndexNotFoundException e1) {
-                if (log.isDebugEnabled()) {
-                    log.debug("No such indices for pattern {}, use raw value", localRequestedPatterns);
-                }
+        private void resolveToLocalAll() {
+            aliases.add(Resolved.ANY);
+            allIndices.add(Resolved.ANY);
+            originalRequested.add(Resolved.ANY);
+        }
 
-                matchingAllIndices = dateResolvedLocalRequestedPatterns;
-            }
+        private void resolveTo(Iterable<String> matchingAliases, Iterable<String> matchingAllIndices, String[] original, Iterable<String> remoteIndices) {
+            aliases.addAll(matchingAliases);
+            allIndices.addAll(matchingAllIndices);
+            originalRequested.add(original);
+            this.remoteIndices.addAll(remoteIndices);
         }
 
-        return new Resolved(matchingAliases, matchingAllIndices, Arrays.asList(requestedPatterns0), remoteIndices);
+        @Override
+        public String[] provide(String[] original, Object localRequest, boolean supportsReplace) {
+            final IndicesOptions indicesOptions = indicesOptionsFrom(localRequest);
+            final boolean enableCrossClusterResolution = localRequest instanceof FieldCapabilitiesRequest || localRequest instanceof SearchRequest;
+            // skip the whole thing if we have seen this exact resolveIndexPatterns request
+            if (alreadyResolved.add(new MultiKey(indicesOptions, enableCrossClusterResolution, original))) {
```

Comment:
Do keys have correctly overridden equals (original - Arrays.equals(original, that.original) was done in previous code) ? Is there a way to simply to avoid complex structures like MultiKey in Set ?

Reply:
Good catch. Wrapped `original` in `MultiKey` for proper Array comparison.